### PR TITLE
Use new PHP7.2 functions in hasColorSupport

### PR DESCRIPTION
**Pull request by johnstevenson**
| Q             | A
| ------------- | ---
| Branch?       |  2.7
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no 
| Tests pass?   | yes 
| Fixed tickets | 
| License       | MIT
| Doc PR        | 

Fixes bc break in #26609 
Reference: https://github.com/composer/xdebug-handler/blob/master/src/Process.php#L111

### CODE REVIEW

**Review by nicolas-grekas — CHANGES_REQUESTED** [review]
There are more places that check for color support (`grep ConEmuANSI src/ -r`)
could you please sync them all?

- src/Symfony/Bridge/PhpUnit/DeprecationErrorHandler.php
- src/Symfony/Component/VarDumper/Dumper/CliDumper.php

**Comment by johnstevenson** [comment]
Sure. But I could do with some info regarding CliDumper:

The constructor checks for Windows color support (so it can set a restricted style set), but only uses environment variables and not the output.

1. I guess it is safe to use `$this->outputStream` here?
2. Or would it be better to set the styles within the `supportsColors` method later?

The `supportsColors` method does one type of check if `$this->outputStream` is not the default (`is_resource` && `posix_isatty`, which doesn't include windows) and a different, more stringent one if it is. An explanation why would be really useful. Likewise why if it is not the default, the method returns before checking for color options in `argv`.

I was wondering if the output needs to be differentiated, and if so, then what are the criteria for each type.

**Comment by nicolas-grekas** [comment]
Which color set does new versions of Windows support? If it doesn't need the reduced set, then the current logic in the CliDumper constructor should be fine, so that only supportsColors should be updated.

**Comment by johnstevenson** [comment]
Windows 10 has supported 24-bit color for over a year now, but you can obviously still run PHP7.2 on earlier versions.

File: src/Symfony/Component/Console/Output/StreamOutput.php [hunk 1]
```diff
@@ -81,31 +81,29 @@ protected function doWrite($message, $newline)
...
+                && @sapi_windows_vt100_support($this->stream))
                 || false !== getenv('ANSICON')
                 || 'ON' === getenv('ConEmuANSI')
                 || 'xterm' === getenv('TERM');
         }
 
-        return function_exists('posix_isatty') && @posix_isatty($this->stream);
+        if (function_exists('stream_isatty')) {
+            return @stream_isatty($this->stream);
+        } elseif (function_exists('posix_isatty')) {
```
Comment (stof) [inline src/Symfony/Component/Console/Output/StreamOutput.php]: just `if`, not `elseif` as the previous case returns

Comment (johnstevenson) [reply src/Symfony/Component/Console/Output/StreamOutput.php]: Okay

**Comment by johnstevenson** [comment]
I've synced DeprecationErrorHandler.php, but without any info about CliDumper.php I have no idea if my changes are correct.

I've kept the current logic in the CliDumper constructor, which is not ideal since it won't show 24-bit color output in cmd.exe on Windows 10. Incidentally, the Windows tests pass for me locally, but not on Appveyor for some reason:

```
There were 3 failures:
1) Symfony\Component\HttpKernel\Tests\DataCollector\DumpDataCollectorTest::testCollectDefault
Failed asserting that two strings are identical.
--- Expected
+++ Actual
@@ @@
-DumpDataCollectorTest.php on line 65:
-123
+DumpDataCollectorTest.php on line 65:
+123

C:\projects\symfony\src\Symfony\Component\HttpKernel\Tests\DataCollector\DumpDataCollectorTest.php:73

2) Symfony\Component\HttpKernel\Tests\DataCollector\DumpDataCollectorTest::testFlush
Failed asserting that two strings are identical.
--- Expected
+++ Actual
@@ @@
-DumpDataCollectorTest.php on line 123:
-456
+DumpDataCollectorTest.php on line 123:
+456

C:\projects\symfony\src\Symfony\Component\HttpKernel\Tests\DataCollector\DumpDataCollectorTest.php:129

3) Symfony\Component\HttpKernel\Tests\DataCollector\DumpDataCollectorTest::testFlushNothingWhenDataDumperIsProvided
Failed asserting that two strings are identical.
--- Expected
+++ Actual
@@ @@
-DumpDataCollectorTest.php on line 142:
-456
+DumpDataCollectorTest.php on line 142:
+456

C:\projects\symfony\src\Symfony\Component\HttpKernel\Tests\DataCollector
```
I've also added references to the source of the code, but I'll remove them if not considered desirable.

File: src/Symfony/Bridge/PhpUnit/DeprecationErrorHandler.php [hunk 2]
```diff
@@ -178,17 +178,38 @@ public static function register($mode = false)
         }
     }
 
+    /**
+     * Returns true if STDOUT is defined and supports colorization.
+     *
+     * Reference: Composer\XdebugHandler\Process::supportsColor
+     * https://github.com/composer/xdebug-handler
+     *
+     * @return bool
```
Comment (nicolas-grekas) [inline src/Symfony/Bridge/PhpUnit/DeprecationErrorHandler.php]: this annotation should be removed and replaced by a return-type on the hasColorsSupport (same in other files below)

**Review by nicolas-grekas — COMMENTED** [review]
LGTM, with one minor comment.
About the CliDumper constructor, we have to use 16 colors only on some terminals, which is what we do by default. Do you think we can improve the logic and somehow detect when the native Windows terminal supports 24b colors? That's what missing there.

File: src/Symfony/Bridge/PhpUnit/DeprecationErrorHandler.php (hunk 2 above)
Comment (johnstevenson) [reply src/Symfony/Bridge/PhpUnit/DeprecationErrorHandler.php]: ?? Sorry, you've lost me!

**Comment by johnstevenson** [comment]
> Do you think we can improve the logic and somehow detect when the native Windows terminal supports 24b colors?

Probably, but I need you to answer my original question:

The constructor checks for Windows color support (so it can set a restricted style set), but only uses environment variables and not the output.

   1. I guess it is safe to use $this->outputStream here?
   2. Or would it be better to set the styles within the supportsColors method later?

File: src/Symfony/Bridge/PhpUnit/DeprecationErrorHandler.php (hunk 2 above)
Comment (nicolas-grekas) [reply src/Symfony/Bridge/PhpUnit/DeprecationErrorHandler.php]: Oups sorry, branch 2.7 which doesn't support scalar type hints.

**Comment by nicolas-grekas** [comment]
It is not safe to use `this->outputStream`. The check should be done against a global stream `STDOUT` basically.

File: src/Symfony/Component/VarDumper/Dumper/CliDumper.php [hunk 3]
```diff
@@ -477,4 +464,69 @@ protected function dumpLine($depth, $endOfValue = false)
...
+     *
+     * Note that this does not check an output stream, but relies on environment
+     * variables from known implementations, or a PHP and Windows version that
+     * supports true color.
+     *
+     * @return bool
+     */
+    private function isWindowsTrueColor()
+    {
+        $result = strval(getenv('ANSICON_VER')) >= '183'
```
Comment (nicolas-grekas) [inline src/Symfony/Component/VarDumper/Dumper/CliDumper.php]: thanks to PHP type juggling, this should be simplified to `183 <= getenv('ANSICON_VER')`

**Review by nicolas-grekas — APPROVED** [review]
(with one minor comment)

File: src/Symfony/Component/VarDumper/Dumper/CliDumper.php (hunk 3 above)
Comment (johnstevenson) [reply src/Symfony/Component/VarDumper/Dumper/CliDumper.php]: Ah yes, that is a much better way of doing it.

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @johnstevenson.

File: src/Symfony/Component/VarDumper/Dumper/CliDumper.php (hunk 3 above)
Comment (fabpot) [reply src/Symfony/Component/VarDumper/Dumper/CliDumper.php]: fixed in 923417122a

**Comment by johnstevenson** [comment]
Thanks for fixing this remaining item, and apologies for not fixing it myself (life/time problem!)

**Pull request merged**